students/MicahBraun/Lesson 06/test_mailroom_functions.py

```diff
@@ -0,0 +1,73 @@
+# ----------------------------------------------------------------------------------------------------------------------
+# NAME: MICAH BRAUN
+# PROJECT: Lesson06: test_mailroom_functions.py
+# PURPOSE: Unit testing mailroom.py (mailroom_refactor.py)
+# DATE: 07/11/2018
+#
+# DESCRIPTION: This test suite was meant to be a means for debugging the mailroom_refactor.py module however the
+#              suite is unsuccessful at debugging the material as I am unable to get the correct arguments to
+#              work with my existing code. I could not find a workable and descriptive solution that mocked user-
+#              -input that would work in my module and so ended up with a bunch of jumbled mess.
+#
+#              Note: mailroom_refactor.py was updated (see notes in file)
+# ----------------------------------------------------------------------------------------------------------------------
+import pytest
+import mock
+from mailroom_refactor import menu, add_donors, add_donation, thankyou, createreport, writeletters, displaylist, \
+    update_list, get_directory, averagedonations
+
+
+class TestMailroom():
+    """
+    Test Suite
+    """
+
+    @pytest.mark.skip(reason="No way of testing this function at present.")
+    def test_menu(self):
+
+        """
+        Test menu behavior
+        """
+        # This all fails... Not sure how to get this to work the way I want it to.
+
+        with mock.patch('mailroom_refactor.menu', return_value="a"):
+            menu() == "Leaving menu..."
+
+        with mock.patch('mailroom_refactor.menu', return_value="b"):
+            assert menu() == "Leaving menu..."
+
+        with mock.patch('mailroom_refactor.menu', return_value="c"):
+            assert menu() == "Leaving menu..."
+
+        with mock.patch('mailroom_refactor.menu', return_value="d"):
+            assert menu() == "Exiting Program...[Press ENTER]"
+
+    @pytest.mark.skip(reason="No way of testing this function at present.")
+    def test_add_donors(self):
+        """
+        Test name addition behavior (allows for int/floats?)
+        """
+
+    @pytest.mark.skip(reason="No way of testing this function at present.")
+    def test_add_donations(self):
+        """
+        Test for donation amounts being added to compiled list
+        """
+
+
+    @pytest.mark.skip(reason="No way of testing this function at present.")
+    @mock.patch('module_under_test.input', create=True)
+    def test_average(self, mocked_input):
+        """
+        Test meant to check that averages are coming out correctly
+        """
+        # NOTE: My tests are all over the place, I am not finding any sources that are comparable
+        #       with my source code to be able to implement a mock/Magicmock to mimic user-input
+        #       short of perhaps re-writing the entire thing...
+        mocked_input.side_effect = [100, 4]
+        result = averagedonations(100, 4)
+        assert result == mocked_input
```

Comment:
I am not seeing a single working unit test here? 

--------------
Please review below working example:

first off, it is best to simplify your data structure. consider this actually being returned from your database, it is most likely not going to have the extra header for example that you have there... the simpler it is the simpler the code that interacts with it.


```
donor_dict = {
              'Rudolph Soares': [1335, 3, 445],
              'Josef Mistretta': [7500, 4, 1875],
              'Joye Agostini': [600, 2, 300],
              'Rachelle Levan': [12100, 5, 2420],
              'Vena Ussery': [4000, 8, 500],
              'Efrain Lager': [795, 9, 88.34],
              'Mee Heine': [4600, 4, 1150],
              'Tanya Essex': [75000, 2, 37500]}
```

here is my version of write_letters, as you can see interaction with donor_dict is extremely simple; i don't even really care about the directory.. but that can be added if really desired. you could have a wrapper function that determines desired user directory then calls this function. what this provides you is a testable code. 
```
def write_letters(directory=None):
    template = ('{date}\n\nDear {first_name}, \n\nThank you for your continued support through your contribution '
        'of {donation} towards our Foundation\'s fundraising goal.\n\nBest wishes,\n'
        'Foundation Board of Directors\n')
    for donor_name, donor_info in donor_dict.items():
        filename = "%s.txt" % donor_name
        with open(filename, 'w') as fh:
            fh.write(template.format(date=today, first_name=donor_name, donation=locale.currency(donor_info[0], grouping=True)))
```
here is very simple example of what your unit test code could look like; again very simple, no mocks.

it could be extended to also assert contents of the file if desired.
```

def test_write_letters():
    write_letters()

    for donor in donor_dict:
        assert os.path.isfile("%s.txt" % donor_name)
```

----------------
there is more mailroom code that can be unit tested but it would of course require some extensive refactor.

Reply:
@mcbraunie here are some other components of your program that should be testable with simple unit tests like the one from Natasha's example:

- the ability to add and retrieve specific donors from your data structure
- the ability to create a donor report with the correct donor data
- the ability to create thank you letters customized for a particular donor

Your program accomplishes these things, but you will need to refactor your code to make it testable and this will also have the result of making your code more flexible, with tightly focused functions that return testable values. 

A concrete example would be to write your functions so that they return the donor report and thank you letter as strings. You can then write unit tests that assert these strings contain the correct content based on the existing data in your donor data structure.